webapi/src/main/scala/org/knora/webapi/messages/util/search/gravsearch/types/InferringGravsearchTypeInspector.scala

```diff
@@ -793,14 +836,149 @@ class InferringGravsearchTypeInspector(nextInspector: Option[GravsearchTypeInspe
-      * @param usageIndex         an index of entity usage in the query.
-      * @return a new intermediate result.
-      */
+     * Gets the iri of the type information.
+     *
+     * @param typeInfo    a GravsearchEntityTypeInfo.
+     * @return the IRI of the typeInfo as a [[SmartIri]].
+     **/
+    def iriOfGravsearchTypeInfo(typeInfo: GravsearchEntityTypeInfo): SmartIri = {
+        typeInfo match {
+            case propertyTypeInfo: PropertyTypeInfo => propertyTypeInfo.objectTypeIri
+            case nonPropertyTypeInfo: NonPropertyTypeInfo => nonPropertyTypeInfo.typeIri
+            case _ => throw GravsearchException(s"There is an invalid type")
+        }
+    }
+
+    /**
+     * Sanitize determined results. If there were multiple resource types, replace with common base class
+     *
+     * @param lastResults this type inspection results.
+     * @param querySchema the ontology schema that the query is written in.
+     **/
+    def sanitizeInconsistentResourceTypes(lastResults: IntermediateTypeInspectionResult, querySchema: ApiV2Schema, entityInfo: EntityInfoGetResponseV2): IntermediateTypeInspectionResult = {
+
+        //This method gets the isResourceType flag of GravsearchEntityTypeInfo
+        def getIsResourceFlags(typeInfo: GravsearchEntityTypeInfo): Boolean = {
+            typeInfo match {
+                case propertyTypeInfo: PropertyTypeInfo => propertyTypeInfo.objectIsResourceType
+                case nonPropertyTypeInfo: NonPropertyTypeInfo=> nonPropertyTypeInfo.isResourceType
+                case _ => throw GravsearchException(s"There is an invalid type")
+            }
+        }
+
+        // This method finds the common base class.
+        def findCommonBaseClass(typesToBeChecked: Set[GravsearchEntityTypeInfo]): SmartIri = {
+            val baseClassesOfFirstType: Set[SmartIri] = entityInfo.classInfoMap.get(iriOfGravsearchTypeInfo(typesToBeChecked.head)) match {
+                case Some(classDef: ReadClassInfoV2) =>
+                    classDef.allBaseClasses
+                case _ => Set.empty[SmartIri]
+            }
+            if (baseClassesOfFirstType.nonEmpty) {
+                val commonBaseClasses: Set[SmartIri] = typesToBeChecked.tail.foldLeft(baseClassesOfFirstType) {
+                    (acc, aType) =>
+                        //get class info of the type Iri
+                        val baseClassesOfType: Set[SmartIri] = entityInfo.classInfoMap.get(iriOfGravsearchTypeInfo(aType)) match {
+                            case Some(classDef: ReadClassInfoV2) =>
+                                classDef.allBaseClasses
+                            case _ => Set.empty[SmartIri]
+                        }
+                        //find the common base classes
+                        baseClassesOfType.intersect(acc)
+                }
+                if (commonBaseClasses.nonEmpty) {
+                    // since the base classes are sorted, first one would be the most specific one.
```

Comment:
`commonBaseClasses` is a `Set`, and a `Set` can't be sorted because it has no order.

Reply:
The classDef.allBaseClasses has an order of base classes, i.e. the most specific one is first. When I do the intersect, it keeps the order.

Reply:
Per our discussion: you're right, but we'll make it a `Seq` in another PR so we can rely on the order.